// lol-js/lib/api/lolStaticData.js
// Generated by CoffeeScript 1.9.0
(function() {
  var api, ld, makeUrl, promiseToCb;

  ld = require('lodash');

  promiseToCb = require('../utils').promiseToCb;

  api = exports.api = {
    fullname: "lol-static-data-v1.2",
    name: "lol-static-data",
    version: "v1.2"
  };

  makeUrl = function(region) {
    return "https://global.api.pvp.net/api/lol/static-data/" + region + "/v1.2";
  };

  exports.methods = {
    getChampionsAsync: function(options) {
      var cacheParams, region, requestParams, _ref, _ref1;
      if (options == null) {
        options = {};
      }
      options = ld.defaults({}, options, {
        region: this.defaultRegion,
        dataById: false
      });
      region = (_ref = options.region) != null ? _ref : this.defaultRegion;
      requestParams = {
        caller: "getChampions",
        region: region,
        url: (makeUrl(region, api)) + "/champion",
        queryParams: ld.pick(options, ['locale', 'version', 'dataById', 'champData']),
        rateLimit: false
      };
      cacheParams = {
        key: (api.fullname + "-champions-" + region + "-" + options.locale + "-" + options.version + "-") + ((options.dataById ? 't' : 'f') + "-" + (((_ref1 = options.champData) != null ? _ref1 : []).join(','))),
        api: api,
        region: region,
        objectType: 'champions',
        params: requestParams.queryParams
      };
      return this._riotRequestWithCache(requestParams, cacheParams, {});
    },
    getChampionByIdAsync: function(id, options) {
      if (options == null) {
        options = {};
      }
      options = ld.extend({}, options, {
        dataById: true
      });
      return this.getChampionsAsync(options).then(function(champions) {
        return champions.data[id];
      });
    },
    getChampionByKeyAsync: function(key, options) {
      if (options == null) {
        options = {};
      }
      options = ld.extend({}, options, {
        dataById: false
      });
      return this.getChampionsAsync(options).then(function(champions) {
        return champions.data[key];
      });
    },
    getChampionByNameAsync: function(name, options) {
      if (options == null) {
        options = {};
      }
      options = ld.extend({}, options, {
        dataById: false
      });
      return this.getChampionsAsync(options).then(function(champions) {
        var answer, championsByName;
        answer = champions.data[name];
        if (answer == null) {
          championsByName = ld.indexBy(champions.data, function(c) {
            return c.name.toLowerCase().replace(/\W/g, '');
          });
          answer = championsByName[name.toLowerCase().replace(/\W/g, '')];
        }
        return answer;
      });
    },
    getItemsAsync: function(options) {
      var cacheParams, region, requestParams, _ref;
      if (options == null) {
        options = {};
      }
      options = ld.defaults({}, options, {
        region: this.defaultRegion,
        dataById: false
      });
      region = (_ref = options.region) != null ? _ref : this.defaultRegion;
      requestParams = {
        caller: "getItems",
        region: region,
        url: (makeUrl(region, api)) + "/item",
        queryParams: ld.pick(options, ['locale', 'version', 'tags']),
        rateLimit: false
      };
      cacheParams = {
        key: (api.fullname + "-champions-" + region + "-" + options.locale + "-" + options.version + "-") + options.tags.join(","),
        api: api,
        region: region,
        objectType: 'items',
        params: requestParams.queryParams
      };
      return this._riotRequestWithCache(requestParams, cacheParams, {});
    },
    getItemByIdAsync: function(id, options) {
      if (options == null) {
        options = {};
      }
      return this.getItemsAsync(options).then(function(objects) {
        return objects.data[id];
      });
    },
    getVersionsAsync: function(options) {
      var cacheParams, region, requestParams, _ref;
      if (options == null) {
        options = {};
      }
      region = (_ref = options != null ? options.region : void 0) != null ? _ref : this.defaultRegion;
      requestParams = {
        caller: "getVersions",
        region: region,
        url: (makeUrl(region, api)) + "/versions",
        rateLimit: false
      };
      cacheParams = {
        key: api.fullname + "-versions-" + region,
        api: api,
        region: region,
        objectType: 'versions',
        params: {}
      };
      return this._riotRequestWithCache(requestParams, cacheParams, {});
    },
    teamNameToId: function(teamName) {
      if (teamName.toLowerCase() === "blue") {
        return 100;
      } else {
        return 200;
      }
    }
  };

}).call(this);
